src/data/string/ustring.h

```diff
@@ -75,72 +76,121 @@ uchar string_to_uchar(const char* p);
 uchar string_to_uchar(const std::string& s);
 std::string uchar_to_string(uchar uc);
 
+class fallback_base {
+public:
+  virtual ~fallback_base() {
+  }
+
+  virtual uchar fallback(const std::string& bytes,
+			 const char* hint) = 0;
+};
+
+class exception_fallback : public fallback_base {
+public:
+  virtual ~exception_fallback() {
+  }
+
+  virtual uchar fallback(const std::string& bytes,
+			 const char* hint) {
+    throw std::invalid_argument(hint);
+    return 0x0000;
+  }
+};
+
+class replacement_fallback : public fallback_base {
+public:
+  replacement_fallback(uchar replace) : replace(replace) {
+  }
+
+  virtual ~replacement_fallback() {
+  }
+
+  virtual uchar fallback(const std::string& bytes,
+			 const char* hint) {
+    if (bytes.empty()) {
+      return 0x0000;
+    }
+    return replace;
+  }
+
+private:
+  uchar replace;
+};
+
 namespace detail {
 template <class InputIterator1, class InputIterator2>
-uchar chars_to_uchar_impl(InputIterator1& in, InputIterator2 end)
+uchar chars_to_uchar_impl(InputIterator1& in, InputIterator2 end,
+			  fallback_base& fb)
 {
+  InputIterator1 begin = in;
   if (in == end) {
-    throw std::invalid_argument("Invalid UTF-8: UTF-8 byte sequences are empty");
+    return fb.fallback(std::string(begin, in),
```

Comment:
Should `chars_to_uchar` return a value for empty string, or not?

This is not encoding error but invalid iterator usage.
So can throw out_of_range error or length_error even if `replacement_fallback` is used.